---
title: 'Using ITS2 to examine the mycobiome of corals'
author: "James Fifer, Chris Reyes"
date: "February 21, 2019"
output: html_document
---
The symbiont communities present in coral reefs have been widely studied with much research focusing on the diversity of the symbioses across species and environments. By contrast, the fungal communities present in coral reefs are rarely studied and it remains unclear how diverse these communities are. Metabarcoding data used to analyze coral-symbiont relationships provide a readily available resource and opportunity to investigate the mycobiome of coral reefs. In order to investigate the mycobiome of corals, we obtained metabarcoding data from Ali et al.(in press), which examined the diversity of different symbiont assemblages and used ITS2 to characterize Symbiodiniaceae communities in coral recruits and coral adult hosts. We similarly used metabarcoding of the ITS2 locus but mapped the data to a fungal reference database to identify how fungal communities differed between coral recruits and coral adults. Using the dada2 pipeline from Ali et al. (in press) we found  only one class of fungi, Agaricomycetes, to be substantially abundant in our samples. We also found that coral recruits host a more diverse mycobiome than coral adults. We, however, demonstrate these findings are skewed by taxa assignment using a narrow ITS2 database and hope to provide a method of filtering sequences to avoid this bias. 

Version Control:
R version 3.5.1 (2018-07-02)
"dada2" 1.10.1
"shortRead" 1.40.0
"ggplot2" 3.1.0
"phyloseq" 1.26.1

Most of the script is based on the DADA2 1.8 Pipeline:
https://benjjneb.github.io/dada2/tutorial.html
with edits by Carly D. Kenkel & Alizah Ali & Nicola Kriefall & Sarah Davies for the purpose of Ali et al. (in press)

```{r}
###Loading packages and setting up

# source("https://bioconductor.org/biocLite.R")
# biocLite("dada2")
# biocLite('vegan')
#####################################
library(dada2); #packageVersion("dada2"); citation("dada2")
library(ShortRead); #packageVersion("ShortRead")
library(ggplot2); #packageVersion("ggplot2")
library(phyloseq); #packageVersion("phyloseq")
library(data.table)
library(dplyr)

#Set path to unzipped, renamed fastq files
setwd("C:/Users/james/Documents/BOSTON/Ecological & evolutionary genomics/R stuff/")
path <- "C:/Users/james/Documents/BOSTON/Ecological & evolutionary genomics/R stuff/Community_Data/Comm_Data" # CHANGE ME to the directory containing the fastq files after unzipping.
fns <- list.files(path)
#Let's make sure that all of our files are there
fns
```

```{r}
###Trimming and Filtering

fastqs <- fns[grepl(".fastq$", fns)]
fastqs <- sort(fastqs) # Sort ensures reads are in same order
#fnFs <- fastqs[grepl("_R1", fastqs)] # Just the forward read files- these are old 454 data but most data are paired end

# Get sample names, assuming files named as so: SAMPLENAME_XXX.fastq; OTHERWISE MODIFY
sample.names <- sapply(strsplit(fastqs, ".fastq"), `[`, 1) #the last number will select the field for renaming
sample.names
# Specify the full path to the fnFs
fnFs <- file.path(path, fastqs)
fnFs


#########Visualize Raw data

#First, lets look at quality profile of R1 reads
plotQualityProfile(fnFs[c(1,2,3,4,5,6,7,8,9)])
plotQualityProfile(fnFs[c(10,11,12,13,14,15,16,17,18)])
plotQualityProfile(fnFs[c(19,20,21,22,23,24,25,26,27)])
plotQualityProfile(fnFs[c(28,29,30,31,32,33,34,35)])
plotQualityProfile(fnFs[c(36,37,38,39,40,41)])

#Recommend trimming where quality profile crashes - in this case, forward reads mostly fine up to 300
#For common ITS amplicon strategies with paired end reads, it is undesirable to truncate reads to a fixed length due to the large amount of length variation at that locus. That is OK, just leave out truncLen. Make sure you removed the forward and reverse primers from both the forward and reverse reads though! 

#truncLen=c(200,200), #leaves ~50bp overlap
#Make directory and filenames for the filtered fastqs
filt_path <- file.path(path, "trimmed")
if(!file_test("-d", filt_path)) dir.create(filt_path)
filtFs <- file.path(filt_path, paste0(sample.names, "_F_filt.fastq.gz"))
?filterAndTrim
# Filter
out <- filterAndTrim(fnFs, filtFs, truncLen= 300, #end of single end reads = approx. 300 bp
                     maxN=0, #DADA does not allow Ns
                     maxEE=1, #allow 1 expected errors, where EE = sum(10^(-Q/10)); more conservative, model converges
                     truncQ=2, 
                     trimLeft=20, #N nucleotides to remove from the start of each read: ITS2 primer = F 20bp
                     rm.phix=TRUE, #remove reads matching phiX genome
                     compress=TRUE, multithread=FALSE) # On Windows set multithread=FALSE

head(out)
tail(out)

#A word on Expected Errors vs a blanket quality threshold
#Take a simple example: a read of length two with quality scores Q3 and Q40, corresponding to error probabilities P=0.5 and P=0.0001. The base with Q3 is much more likely to have an error than the base with Q40 (0.5/0.0001 = 5,000 times more likely), so we can ignore the Q40 base to a good approximation. Consider a large sample of reads with (Q3, Q40), then approximately half of them will have an error (because of the P=0.5 from the Q2 base). We express this by saying that the expected number of errors in a read with quality scores (Q3, Q40) is 0.5.
#As this example shows, low Q scores (high error probabilities) dominate expected errors, but this information is lost by averaging if low Qs appear in a read with mostly high Q scores. This explains why expected errors is a much better indicator of read accuracy than average Q.
```

```{r}
###Learn error rates

#DADA2 learns its error model from the data itself by alternating estimation of the error rates and the composition of the sample until they converge on a jointly consistent solution (this is similar to the E-M algorithm)
#As in many optimization problems, the algorithm must begin with an initial guess, for which the maximum possible error rates in this data are used (the error rates if only the most abundant sequence is correct and all the rest are errors).

setDadaOpt(MAX_CONSIST=30) #increase number of cycles to allow convergence
errF <- learnErrors(filtFs, multithread=TRUE)
#Maximum cycles was set to 30, but Convergence was found after 4 rounds

#sanity check: visualize estimated error rates
#error rates should decline with increasing qual score
#red line is based on definition of quality score alone
#black line is estimated error rate after convergence
#dots are observed error rate for each quality score

plotErrors(errF, nominalQ=TRUE) 
```

```{r}
###Dereplicate reads

#Dereplication combines all identical sequencing reads into “unique sequences” with a corresponding “abundance”: the number of reads with that unique sequence. 
#Dereplication substantially reduces computation time by eliminating redundant comparisons.
#DADA2 retains a summary of the quality information associated with each unique sequence. The consensus quality profile of a unique sequence is the average of the positional qualities from the dereplicated reads. These quality profiles inform the error model of the subsequent denoising step, significantly increasing DADA2’s accuracy.
derepFs <- derepFastq(filtFs, verbose=TRUE)
# Name the derep-class objects by the sample names
names(derepFs) <- sample.names
```

```{r}
###Infer sequence variants

#Must change some of the DADA options b/c original program optomized for ribosomal data, not ITS - from github, "We currently recommend BAND_SIZE=32 for ITS data." leave as default for 16S/18S
setDadaOpt(BAND_SIZE=32)
dadaFs <- dada(derepFs, err=errF, multithread=TRUE)

#now, look at the dada class objects by sample
#will tell how many 'real' variants in unique input seqs
#By default, the dada function processes each sample independently, but pooled processing is available with pool=TRUE and that may give better results for low sampling depths at the cost of increased computation time. See our discussion about pooling samples for sample inference. 
dadaFs[[1]]
dadaFs[[25]]

#construct sequence table
seqtab <- makeSequenceTable(dadaFs)
head(seqtab)
#
plot(table(nchar(getSequences(seqtab))))
```

```{r}
###Remove chimeras

#The core dada method removes substitution and indel errors, but chimeras remain. 
#Fortunately, the accuracy of the sequences after denoising makes identifying chimeras easier 
#than it is when dealing with fuzzy OTUs: all sequences which can be exactly reconstructed as 
#a bimera (two-parent chimera) from more abundant sequences.

seqtab.nochim <- removeBimeraDenovo(seqtab, method="consensus", multithread=TRUE, verbose=TRUE)
dim(seqtab.nochim)
# Identified 1 bimeras out of 117 input sequences.

sum(seqtab.nochim)/sum(seqtab)
#The fraction of chimeras varies based on factors including experimental procedures and sample complexity, 
#Most of your reads should remain after chimera removal (it is not uncommon for a majority of sequence variants to be removed though)
#For our sample, this ratio was 0.9998201, there was only 1 bimera

write.csv(seqtab,file="Alizah_seqtab.csv")
write.csv(seqtab.nochim,file="Alizah_nochim.csv")
```

```{r}
###Track read stats

getN <- function(x) sum(getUniques(x))
track <- cbind(out, sapply(dadaFs, getN), sapply(dadaFs, getN), rowSums(seqtab), rowSums(seqtab.nochim))
colnames(track) <- c("input", "filtered", "denoised", "merged", "tabled", "nonchim")
rownames(track) <- sample.names
head(track)
tail(track)

write.csv(track,file="ReadFilterStats_AllData_final.csv",row.names=TRUE,quote=FALSE)
```

We modified the script to examine the fungal communities from our ITS2 dataset. We used the  database UNITE plus INSD https://dx.doi.org/10.15156/BIO/786347. This database offers a curated version of public fungal ITS sequences (Nilsson et al., 2018) and formatting is compatible with the dada2 pipeline.  We ran phyloseq with a minimum bootstrap of 50 (per guidelines of Wang et al., 2007).  

```{r}
###Assign Taxonomy

#It is common at this point, especially in 16S/18S/ITS amplicon sequencing, to classify sequence variants taxonomically. 
#DADA2 provides a native implementation of the RDP's naive Bayesian classifier. The assignTaxonomy function takes a set of sequences and a training set of taxonomically classified sequences, and outputs the taxonomic assignments with at least minBoot bootstrap confidence.
#Here, I have supplied a modified version of the GeoSymbio ITS2 database listing more taxonomic info as phyloseq requires (Franklin et al. 2012)
#For example: GeoSymbio data (taken from "all clades" at https://sites.google.com/site/geosymbio/downloads):
#>A1.1
#modified version for phyloseq looks like this instead:
#>Symbiodinium; Clade A; A1.1

fungitaxa<- assignTaxonomy(seqtab.nochim, "C:/Users/james/Documents/BOSTON/Ecological & evolutionary genomics/R stuff_Assignment 1/Community_Data/sh_general_release_dynamic_02.02.2019.fasta", minBoot=50,multithread=TRUE,tryRC=TRUE,outputBootstraps=FALSE)
setwd("C:/Users/james/Documents/BOSTON/Ecological & evolutionary genomics/R stuff_Assignment 1/")
unname(head(fungitaxa, 30))
unname(fungitaxa)
#Now, save outputs so can come back to the analysis stage at a later point if desired
saveRDS(seqtab.nochim, file="final_seqtab_nochim.rds")
saveRDS(fungitaxa, file="final_fungitaxa_blastCorrected.rds")
#If you need to read in previously saved datafiles
seqtab.nochim <- readRDS("final_seqtab_nochim.rds")
fungitaxa <- readRDS("final_fungitaxa_blastCorrected.rds")
```

```{r }
###Phyloseq

#import dataframe holding sample information
#have your samples in the same order as the seqtab file in the rows, variables as columns
setwd("C:/Users/james/Documents/BOSTON/Ecological & evolutionary genomics/R stuff_Assignment 1/Community_Data/")
samdf<-read.csv("variabletable.csv")
head(samdf)
head(seqtab.nochim)
head(taxa)
rownames(samdf) <- samdf$sample

# Construct phyloseq object (straightforward from dada2 outputs)
ps <- phyloseq(otu_table(seqtab.nochim, taxa_are_rows=FALSE), 
               sample_data(samdf), 
               tax_table(fungitaxa))

#Bar-plots
top90 <- names(sort(taxa_sums(ps), decreasing=TRUE))[1:130]
ps.top90 <- transform_sample_counts(ps, function(OTU) OTU/sum(OTU))
ps.top90 <- prune_taxa(top90, ps.top90)
plot_bar(ps.top90, x="Sample", fill="Class")
```

We noticed that there were a lot of NAs in our plot. We can examine diversity between our treatments which will take into account OTUs not assigned. We hypothesized that the NAs are all potentially fungi.

```{r}
#Visualize alpha-diversity - ***Should be done on raw, untrimmed dataset***
#total species diversity in a landscape (gamma diversity) is determined by two different things, the mean species diversity in sites or habitats at a more local scale (alpha diversity) and the differentiation among those habitats (beta diversity)
#Shannon:Shannon entropy quantifies the uncertainty (entropy or degree of surprise) associated with correctly predicting which letter will be the next in a diverse string. Based on the weighted geometric mean of the proportional abundances of the types, and equals the logarithm of true diversity. When all types in the dataset of interest are equally common, the Shannon index hence takes the value ln(actual # of types). The more unequal the abundances of the types, the smaller the corresponding Shannon entropy. If practically all abundance is concentrated to one type, and the other types are very rare (even if there are many of them), Shannon entropy approaches zero. When there is only one type in the dataset, Shannon entropy exactly equals zero (there is no uncertainty in predicting the type of the next randomly chosen entity).
#Simpson:equals the probability that two entities taken at random from the dataset of interest represent the same type. equal to the weighted arithmetic mean of the proportional abundances pi of the types of interest, with the proportional abundances themselves being used as the weights. Since mean proportional abundance of the types increases with decreasing number of types and increasing abundance of the most abundant type, λ obtains small values in datasets of high diversity and large values in datasets of low diversity. This is counterintuitive behavior for a diversity index, so often such transformations of λ that increase with increasing diversity have been used instead. The most popular of such indices have been the inverse Simpson index (1/λ) and the Gini–Simpson index (1 − λ).

plot_richness(ps, x="site", measures=c("Shannon", "Simpson"), color="in_off") + theme_bw()
```

Diversity analysis shows differences between treatments, but when we blast these sequences against genbank or another ITS2-specific database like Geosymbio for example, we see that many of these sequences are actually Symbiodiniaceae with high confidence. 

```{r, engine = 'bash', eval = FALSE}
#create file from fungitaxa csv of just sequences
 awk '{print NR, $0}' OFS="\t" fungitaxa.csv > taxa.seq.tab
#assign numbers to sequences and transform into fasta format
 awk '{print ">"$1"\n"$2}' taxa.seq.tab > seqs.fa
#blast against desired database (recommend NCBI nt), in this case we did local GeoSymbio
#set e value cut off to very high confidence < e-40
makeblastdb -in GeoSymbio_ITS2_LocalDatabase_verForPhyloseq.fasta -dbtype nucl
blastn -query seqs.fa -db GeoSymbio_ITS2_LocalDatabase_verForPhyloseq.fasta -evalue 1e-40 -outfmt 6 -max_target_seqs 1 -max_hsps 1 > sym.output
#grep -w -A 2 -f  symID.txt seqs.fa --no-group-separator >symseqoverlap.txt
```


```{r}
###Transform the sequence fasta file into a two column matrix so that it is easier to work with and identify the sequence numbers that did not match to the GeoSym database 
sym = read.delim("sym.output", header=FALSE)
dat = sym$V1
missing = setdiff(1:122, dat)
s = read.delim("seqs.fa", header=FALSE)
s.v = as.vector(s[,1])
s.odd = s.v[c(TRUE,FALSE)]
s.odd = sapply(strsplit(s.odd, split='>', fixed=TRUE), function(x) (x[2]))
s.even = s.v[c(FALSE,TRUE)]
s.new = as.data.frame(cbind(s.odd,s.even))

library(dplyr)
# Turn list of missing sequence numbers into a column
s.missing<-data.frame(missing)
#filter from the list of all sequences to obtain sequences that did not match
#to sym with high confidence
notsym.sequences<-s.new %>%
  filter(s.odd %in% s.missing$missing)


write.csv(notsym.sequences, file="notsym.sequences.csv",row.names=FALSE, col.names=TRUE, quote=FALSE)
```
One option would be to rerun the analysis (re-assign this subset against the UNITE database).
However, given that the "notsym" output returns only 19 sequences, we decided to instead accept the Symbiodiniaceae hits as more confident than the fungi hits and only use fungi hits for sequences where assignment against symbiont database returns "NA". 

```{r}
###Rerun the pipeline with the sym database, starting at assigning taxonomy, with a minimum bootstrap of at least 50

symtaxa<- assignTaxonomy(seqtab.nochim, "C:/Users/james/Documents/BOSTON/Ecological & evolutionary genomics/R stuff_Assignment 1/Community_Data/GeoSymbio_ITS2_LocalDatabase_verForPhyloseq.fasta", minBoot=50,multithread=TRUE,tryRC=TRUE,outputBootstraps=FALSE)
#Save if you want to come back later
saveRDS(symtaxa, file="final_symtaxa_blastCorrected.rds")

#If you need to read in previously saved datafiles
seqtab.nochim <- readRDS("final_seqtab_nochim.rds")
symtaxa <- readRDS("final_symtaxa_blastCorrected.rds")

setwd("C:/Users/james/Documents/BOSTON/Ecological & evolutionary genomics/R stuff_Assignment 1/Community_Data/")
samdf<-read.csv("variabletable.csv")
rownames(samdf) <- samdf$sample

# Construct phyloseq object (straightforward from dada2 outputs)
ps <- phyloseq(otu_table(seqtab.nochim, taxa_are_rows=FALSE), 
               sample_data(samdf), 
               tax_table())
# Can look at plots to see where NAs are

top90 <- names(sort(taxa_sums(ps), decreasing=TRUE))[1:130]
ps.top90 <- transform_sample_counts(ps, function(OTU) OTU/sum(OTU))
ps.top90 <- prune_taxa(top90, ps.top90)

plot_bar(ps.top90, x="Sample", fill="Class") 
```

```{r}
#Now we assign NAs as fungi when applicable. 
taxafungi<-read.csv(file="fungitaxa.csv")
taxasym <-read.csv(file="symtaxa.csv")

library(data.table)
setDT(taxafungi); setDT(taxasym)

#replace NAs for col "Class" in taxasym DT with values from col "Class" in taxasym DT and create joint DT
fungisymDF <-taxasym[is.na(Class), Class := taxafungi[copy(.SD), on=.(X), x.Class]]
#save
write.csv(fungisymDF, file="fungisymDF.csv",row.name=TRUE,quote=FALSE)
#now we turn it back into matrix so we can run phyloseq 
fungisymDF<-as.matrix(read.csv(file="fungisymDF.csv",row.names = 2))
#remove extra column that forms because of as.matrix function
fungisymDF<-fungisymDF[,-1]
#sort rows so it matches with seqtab.nochim file
Final_df <- fungisymDF[match(colnames(seqtab.nochim),rownames(fungisymDF)),]

setwd("C:/Users/james/Documents/BOSTON/Ecological & evolutionary genomics/R stuff_Assignment 1/Community_Data/")
samdf<-read.csv("variabletable.csv")
rownames(samdf) <- samdf$sample

# Construct phyloseq object (straightforward from dada2 outputs)
ps <- phyloseq(otu_table(seqtab.nochim, taxa_are_rows=FALSE), 
               sample_data(samdf), 
               tax_table(Final_df))

#Bar-plots
top90 <- names(sort(taxa_sums(ps), decreasing=TRUE))[1:130]
ps.top90 <- transform_sample_counts(ps, function(OTU) OTU/sum(OTU))
ps.top90 <- prune_taxa(top90, ps.top90)

plot_bar(ps.top90, x="Sample", fill="Class") 
```

We attempted to use ITS2 data from a study examining Symbiodiniaceae to investigate the coral mycobiome. We demonstrate the necessity of an additional filtration step not in the current pipeline to prevent incorrect taxa assignment by phyloseq for mycobiome analysis. We also show, in this dataset, unassigned sequences against the Symbiodiniaceae database were also unassigned when using the UNITE fungi database. However, some sequences that were determined to not be Symbiodiniaceae were assigned to the fungi class Agaricomycetes and this was only the case for recruit samples. Future analyses on the coral mycobiome using ITS2 data should focus on determining whether the UNITE database is an appropiate database and if the ITS2 region is a suitable marker to be used for coral mycobiome assignment. 

References:

Alizah Ali, Nicola G Nicola, Lara E Emery, Carly D Kenkel, Mikhail V Matz, Sarah W Davies
bioRxiv 421339; doi: https://doi.org/10.1101/421339

Nilsson RH, Larsson K-H, Taylor AFS, Bengtsson-Palme J, Jeppesen TS, Schigel D, Kennedy P, Picard K, Gl?ckner FO, Tedersoo L, Saar I, K?ljalg U, Abarenkov K. 2018. The UNITE database for molecular identification of fungi: handling dark taxa and parallel taxonomic classifications. Nucleic Acids Research, DOI: 10.1093/nar/gky1022

Wang Q, Garrity GM, Tiedje JM, Cole JR. Naive Bayesian classifier for rapid assignment of rRNA sequences into the new bacterial taxonomy. Appl Environ Microbiol. 2007;73(16):5261-7.
